### ISSUE TITLE
Driver does not compile for kernels newer than Linux 5.15.32

### ISSUE THREAD

**Issue by jrddunbr** [Gateworks/nrc7292#1]
Hello,

Your driver does not compile for kernels newer than 5.15.32. Introduced in kernel 5.16 is the following change:

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v5.17.1&id=bdc7ca008e1f5539e891187032cb2cbbc3decb5e

This change results in this compilation error:

```
/home/jared/nrc7292/package/host/nrc_driver/source/nrc_driver/nrc/nrc-ssp.c: In function 'nrc_ssp_register':
/home/jared/nrc7292/package/host/nrc_driver/source/nrc_driver/nrc/nrc-ssp.c:591:18: error: implicit declaration of function 'spi_busnum_to_master' [-Werror=implicit-function-declaration]
  591 |         master = spi_busnum_to_master(spi_bus_num);
      |                  ^~~~~~~~~~~~~~~~~~~~
```

I'm going to mitigate the issue by using an older kernel, but I really want to be able to use this device with kernels newer than 5.15 in the near future so I'd love to see an updated (and/or mainlined) driver. Thanks!

**Comment by Gateworks** [comment]
Thanks - I'll try to fix this soon. If you want to submit a pull request I would be happy to pull it :)

With regards to getting this driver mainlined - I would love to see that as well. Please hit up the Newracom folk responsible for the driver: https://github.com/newracom/nrc7292_sw_pkg

**Comment by jrddunbr** [comment]
Yeah, the upstream driver also will need to be updated. It did not look like it was going to be a trivial fix based on the code removed from the kernel for the SPI headers.

I actually couldn't get it to insmod on the 5.15.32 kernel either, but that's likely an unrelated issue I encountered (it compiles just fine). Still need to mess with it.

I am not really great at doing C, so I probably won't be able to make that pull request; I would probably need approval to make that contribution from my employer as well. :/

**Comment by Gateworks** [comment]
Jared, please contact sales@gateworks.com and let us know what your project is and how we can better help you

**Comment by jrddunbr** [comment]
Ah, I probably should have made it clear that this is a personal project and not a project for my employer. All messages in this thread are my personal opinions (and projects), and not those of my employer.

---

I'm working to configure a sensor mesh network with (currently) two devices: a single connection to a backbone network and the other test node as a client. The Raspberry Pi has been out of stock for months, so I have to compile all of the software on x86_64 for my two test systems (although my final build will be installed in an aarch64 system, which I want to minimize downtime of).

One test system is emulating my router and has a connection to the internet, the other test system is configured as a client, and once I get those two systems configured and working correctly, I have two embedded sensors that I will be connecting to that network for monitoring environmental sensors. These sensors are installed on a mobile device that generally speaking is around at most 6-8 floors down in a concrete parking garage, so it will be an interesting distance performance test (I have other uses if it turns out that doesn't work, including a p2p wireless network project between two buildings).

The final router system runs on an aarch64 Ten64 device I already have been using onsite, which doesn't have mainline kernel support until 5.16 ,which is primarily why I need the updated kernel (the other reason being better security and ease of use with an updated operating system).

I obviously can make everything else work on my own, but I could really use your driver on a recent kernel. I can also work with upstream and open an issue to update their driver as well, since they are very similar kernel drivers.

I'm already aware that I need to make modifications to the userspace software to remove platform specific code for the Raspberry Pi, and while on the Gateworks website it suggests that you won't support other hardware than your own, I think that the Linux kernel version support of your driver is isolated enough from your hardware to be a worthy investment in general.

Thank you!
